server/py/services/api/runtime_handlers/base.py

```diff
@@ -633,15 +633,20 @@ def _ensure_run_not_stuck_on_non_terminal_state(
                         "possibly it was preempted or evicted. "
                         "Additional details may be available from Kubernetes events."
                     )
-                logger.info(
-                    "Updating run state", run_uid=run_uid, run_state=RunStates.error
-                )
-                run.setdefault("status", {})["state"] = RunStates.error
 
-                run.setdefault("status", {})["reason"] = reason
-                run.setdefault("status", {})["last_update"] = now.isoformat()
-                run.setdefault("status", {})["end_time"] = now.isoformat()
-                db.store_run(db_session, run, run_uid, project)
+                # Check if the run should be retried, and update its status accordingly
+                run_state, message = self._evaluate_run_retry_state(run, reason)
+                logger.info("Updating run state", run_uid=run_uid, run_state=run_state)
+                run_updates = {
+                    "status.state": run_state,
+                    "status.reason": reason,
+                    "status.status_text": message,
+                    "status.last_update": now.isoformat(),
+                    "status.end_time": now.isoformat(),
```

Comment:
this is the responsibility of _update_run_end_time function on sql to do, why do we do it here too?